ouroboros-network/src/Ouroboros/Network/StrictFingerTree.hs

```diff
@@ -0,0 +1,264 @@
+  , ViewR(..)
+  , viewl
+  , viewr
+
+    -- ** Search
+  , SearchResult(..)
+  , search
+
+    -- ** Splitting
+    -- | These functions are special cases of 'search'.
+  , split
+  , takeUntil
+  , dropUntil
+
+    -- * Transformation
+  , reverse
+
+    -- ** Maps
+  , fmap'
+
+  -- Re-export from "Data.FingerTree"
+  , Measured(..)
+  ) where
+
+import           Prelude hiding (null, reverse)
+
+import           Data.FingerTree (Measured (..))
+import qualified Data.FingerTree as FT
+import           Data.Foldable (foldl')
+import           GHC.Generics (Generic)
+
+infixr 5 ><
+infixr 5 <|, :<
+infixl 5 |>, :>
+
+-- | A @newtype@ wrapper around a 'FT.FingerTree', representing a finger tree
+-- that is strict in its values.
+--
+-- This strictness is not enforced at the type level, but rather by the
+-- construction functions in this module. These functions essentially just
+-- wrap the original "Data.FingerTree" functions while forcing the provided
+-- value to WHNF.
+newtype FingerTree v a = SFT (FT.FingerTree v a)
+  deriving (Eq, Ord, Show)
+
+deriving instance Foldable (FingerTree v)
+deriving instance (Measured v a) => Measured v (FingerTree v a)
+
+{-------------------------------------------------------------------------------
+  Construction
+-------------------------------------------------------------------------------}
+
+-- | /O(1)/. The empty sequence.
+empty :: Measured v a => FingerTree v a
+empty = SFT (FT.empty)
+
+-- | /O(1)/. A singleton sequence.
+singleton :: Measured v a => a -> FingerTree v a
+singleton !a = SFT (FT.singleton a)
+
+-- | /O(n)/. Create a sequence from a finite list of elements.
+-- The opposite operation 'toList' is supplied by the 'Foldable' instance.
+fromList :: (Measured v a) => [a] -> FingerTree v a
+fromList !xs = foldl' (|>) (SFT $ FT.empty) xs
+
+-- | /O(1)/. Add an element to the left end of a sequence.
+-- Mnemonic: a triangle with the single element at the pointy end.
+(<|) :: (Measured v a) => a -> FingerTree v a -> FingerTree v a
+(!a) <| SFT ft = SFT (a FT.<| ft)
+
+-- | /O(1)/. Add an element to the right end of a sequence.
+-- Mnemonic: a triangle with the single element at the pointy end.
+(|>) :: (Measured v a) => FingerTree v a -> a -> FingerTree v a
+SFT ft |> (!a) = SFT (ft FT.|> a)
+
+-- | /O(log(min(n1,n2)))/. Concatenate two sequences.
+(><) :: (Measured v a) => FingerTree v a -> FingerTree v a -> FingerTree v a
+SFT ft >< SFT ft' = SFT (ft FT.>< ft')
+
+{-------------------------------------------------------------------------------
+  Deconstruction
+-------------------------------------------------------------------------------}
+
+-- | /O(1)/. Is this the empty sequence?
+null :: FingerTree v a -> Bool
+null (SFT ft) = FT.null ft
+
+{-------------------------------------------------------------------------------
+  Examining the ends
+-------------------------------------------------------------------------------}
+
+-- | View of the left end of a sequence.
+data ViewL s a
```

Comment:
It's not clear to me why we need `ViewL` and `ViewR`, cannot we just use the originals directly? `fromOriginalViewL` will not be needed.

Reply:
You're right. Thanks.